Fix the kl_div docs

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67443

Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc @brianjo @mruberry

Differential Revision: [D32136888](https://our.internmc.facebook.com/intern/diff/D32136888)

lezcano committed:
Fix the kl_div docs

Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

[ghstack-poisoned]

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

pmeier reviewed (changes requested):
Some style suggestions and fixes inline and one larger possible mixup. Let's check this carefully to not mess this up again.

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

pmeier reviewed (approved):
Only nitpicks and style fixes left.

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

jbschlosser commented on torch/nn/modules/loss.py:
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
-    `Kullback-Leibler divergence`_ is a useful distance measure for continuous
-    distributions and is often useful when performing direct regression over
-    the space of (discretely sampled) continuous output distributions.
+    For tensors of the same shape :math:`y_{\text{model}},\ y_{\text{obs}}`,
```
The math looks correct to me and I appreciate the rigor you approached this with!

Only thing that concerns me is that it's inconsistent with other losses, which generally use the notation `x_n` for input and `y_n` for target. I think it's okay that it's different, but I have a couple suggestions:
* I'd make the connection between the two variables and input / target very explicit (something like `"where y_{model} is the ``input`` and y_{obs} is the target"`)
* Possibly use `y_{true}` / `y_{pred}` instead? this matches TF and I believe fits well with how ML practitioners think (in terms of predicted values vs. ground truth)

Overall I think it's much clearer what's going on :)

BenoitDalFerro replied:
@jbschlosser, suggest actual explicit reference to the P and Q arguments and the implementation being swapped wrt. to math because this is machine learning framework tradition (cf. Tensorflow)

Lezcano replied:
The obs / model came from the probabilistic take on the KL divergence, but I see that using more layman terms might very well be helpful to many.

Also, using this we get rid of parens / em-dashes :D

jbschlosser reviewed (commented):
Thanks for the update! Added a few comments below

pmeier commented on torch/nn/modules/loss.py:
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
-            \operatorname{mean}(L), & \text{if reduction} = \text{`mean';} \\
-            \operatorname{sum}(L),  & \text{if reduction} = \text{`sum'.}
-        \end{cases}
+        if not log_target: # default
+            loss_pointwise = target * (target.log() - input)
+        else:
+            loss_pointwise = target.exp() * (target - input)
```
I feel like Marios version with the default behavior at the top is more readable. In the actual implementation, I would also go for your implementation, but for docs I would value readability higher. Of course this is also subjective.

Lezcano replied:
I'm with @pmeier on this one.

BenoitDalFerro commented on torch/nn/modules/loss.py:
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
+        This differs from the standard mathematical notation :math:`KL(P\ ||\ Q)` where
+        :math:`P` denotes the distribution of the observations and :math:`Q` denotes the model.
+
+    .. note::
+        :attr:`reduction`\ `= "mean"` doesn't return the true kl divergence value, please use
+        :attr:`reduction`\ `= "batchmean"` which aligns with the mathematical definition.
+        In a future release, `"mean"` will be changed to be the same as `"batchmean"`.
```
@jbschlosser, yes absolutely

Lezcano commented on torch/nn/modules/loss.py:
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
 
-    The unreduced (i.e. with :attr:`reduction` set to ``'none'``) loss can be described as:
+    As to avoid underflow issues, to compute this quantity, this loss expects the argument
+    :attr:`input`---:math:`y_{\text{model}}` in the notation above---in the log-space.
```
I'm a big fan of em and en dashes, but I completely see your point. Reverting to parens.

lezcano committed:
Update on "Fix the kl_div docs"


Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc brianjo mruberry

[ghstack-poisoned]

Lezcano commented:
Thanks for the detailed review @jbschlosser ! I just addressed the comments. Could you have a second look at this, please?

jbschlosser reviewed (approved):
LGTM! Thanks for the in-depth improvements :)

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32136888).
